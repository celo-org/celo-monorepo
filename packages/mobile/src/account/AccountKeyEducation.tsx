import { BtnTypes } from '@celo/react-components/components/Button.v2'
import { StackScreenProps } from '@react-navigation/stack'
import React, { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import Education, { EducationTopic, EmbeddedNavBar } from 'src/account/Education'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { Namespaces } from 'src/i18n'
import { accountKey1, accountKey2, accountKey3, accountKey4 } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'

type Props = StackScreenProps<StackParamList, Screens.AccountKeyEducation>

export default function AccountKeyEducation(props: Props) {
  function onComplete() {
    ValoraAnalytics.track(OnboardingEvents.backup_education_complete)
    if (props.route.params?.nextScreen) {
      navigate(props.route.params?.nextScreen)
    } else {
      navigate(Screens.BackupPhrase)
    }
  }

  const { t } = useTranslation(Namespaces.backupKeyFlow6)

  const steps = useSteps()

  useEffect(() => {
    ValoraAnalytics.track(OnboardingEvents.backup_education_start)
  }, [])

  return (
    <Education
      embeddedNavBar={EmbeddedNavBar.Close}
      stepInfo={steps}
      onFinish={onComplete}
      finalButtonText={t('completeEducation')}
      buttonText={t('global:next')}
      finalButtonType={BtnTypes.PRIMARY}
    />
  )
}

function useSteps() {
  const { t } = useTranslation(Namespaces.backupKeyFlow6)
  return React.useMemo(
    () =>
      [
        { image: accountKey1, topic: EducationTopic.backup },
        { image: accountKey2, topic: EducationTopic.backup },
        { image: accountKey3, topic: EducationTopic.backup },
        { image: accountKey4, topic: EducationTopic.backup },
      ].map((step, index) => {
        return {
          ...step,
          title: t(`guide.${index}.title`),
          text: t(`guide.${index}.text`),
        }
      }),
    []
  )
}
