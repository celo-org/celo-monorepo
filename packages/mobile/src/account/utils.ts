import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
