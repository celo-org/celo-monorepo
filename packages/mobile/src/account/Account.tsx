import Link from '@celo/react-components/components/Link'
import SmallButton from '@celo/react-components/components/SmallButton'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import { anonymizedPhone, isE164Number } from '@celo/utils/src/phoneNumbers'
import * as React from 'react'
import { Trans, WithNamespaces, withNamespaces } from 'react-i18next'
import { Clipboard, ScrollView, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import DeviceInfo from 'react-native-device-info'
import SafeAreaView from 'react-native-safe-area-view'
import { Sentry } from 'react-native-sentry'
import { connect } from 'react-redux'
import { devModeTriggerClicked } from 'src/account/actions'
import SettingsItem from 'src/account/SettingsItem'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { resetAppOpenedState, setAnalyticsEnabled, setNumberVerified } from 'src/app/actions'
import { AvatarSelf } from 'src/components/AvatarSelf'
import { FAQ_LINK, TOS_LINK } from 'src/config'
import { features } from 'src/flags'
import { Namespaces } from 'src/i18n'
import { revokeVerification } from 'src/identity/actions'
import { isPhoneNumberVerified } from 'src/identity/verification'
import { headerWithBackButton } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import { navigateToURI, navigateToVerifierApp } from 'src/utils/linking'
import Logger from 'src/utils/Logger'

interface DispatchProps {
  revokeVerification: typeof revokeVerification
  setNumberVerified: typeof setNumberVerified
  resetAppOpenedState: typeof resetAppOpenedState
  setAnalyticsEnabled: typeof setAnalyticsEnabled
  devModeTriggerClicked: typeof devModeTriggerClicked
}

interface StateProps {
  account: string | null
  e164PhoneNumber: string
  devModeActive: boolean
  analyticsEnabled: boolean
}

type Props = StateProps & DispatchProps & WithNamespaces

interface State {
  verified: boolean | undefined
}

const mapStateToProps = (state: RootState): StateProps => {
  return {
    account: state.web3.account,
    devModeActive: state.account.devModeActive || false,
    e164PhoneNumber: state.account.e164PhoneNumber,
    analyticsEnabled: state.app.analyticsEnabled,
  }
}

const mapDispatchToProps = {
  revokeVerification,
  setNumberVerified,
  resetAppOpenedState,
  setAnalyticsEnabled,
  devModeTriggerClicked,
}

export class Account extends React.Component<Props, State> {
  static navigationOptions = headerWithBackButton

  state: State = {
    verified: undefined,
  }

  async componentDidMount() {
    const phoneNumber = this.props.e164PhoneNumber
    const verified = await isPhoneNumberVerified(phoneNumber)
    this.setState({ verified })
  }

  goToProfile = () => {
    CeloAnalytics.track(CustomEventNames.edit_profile)
    navigate(Screens.Profile)
  }

  backupScreen() {
    navigate(Screens.Backup)
  }

  goToInvite() {
    navigate(Screens.Invite)
  }

  goToLanguageSetting() {
    navigate(Screens.Language, { nextScreen: Screens.Account })
  }

  goToLocalCurrencySetting() {
    navigate(Screens.SelectLocalCurrency)
  }

  goToLicenses() {
    navigate(Screens.Licenses)
  }

  goToAnalytics() {
    navigate(Screens.Analytics, { nextScreen: Screens.Account })
  }

  goToFAQ() {
    navigateToURI(FAQ_LINK)
  }

  goToTerms() {
    navigateToURI(TOS_LINK)
  }

  resetAppOpenedState = () => {
    this.props.resetAppOpenedState()
    Logger.showMessage('App onboarding state reset.')
  }

  revokeNumberVerification = async () => {
    if (!isE164Number(this.props.e164PhoneNumber)) {
      Logger.showMessage('Cannot revoke verificaton: number invalid')
      return
    }
    Logger.showMessage(`Revoking verification`)
    this.props.revokeVerification()
  }

  showDebugScreen = async () => {
    navigate(Screens.Debug)
  }

  sendLogsToSupport = async () => {
    if (this.props.e164PhoneNumber) {
      await Logger.emailLogsToSupport(anonymizedPhone(this.props.e164PhoneNumber))
    }
  }

  onPressAddress = () => {
    const { account, t } = this.props
    if (!account) {
      return
    }
    Clipboard.setString(account)
    Logger.showMessage(t('addressCopied'))
  }

  onPressAvatar = () => {
    this.props.devModeTriggerClicked()
  }

  getDevSettingsComp() {
    const { devModeActive } = this.props
    const { verified } = this.state

    if (!devModeActive) {
      return null
    } else {
      return (
        <View style={style.devSettings}>
          <View style={style.devSettingsItem}>
            <Text>Dev Settings</Text>
            <View>
              {verified === undefined && <Text>Checking Verification</Text>}
              {verified === true && <Text>Verified</Text>}
              {verified === false && <Text>Not Verified</Text>}
            </View>
          </View>
          <View style={style.devSettingsItem}>
            <TouchableOpacity onPress={this.revokeNumberVerification}>
              <Text>Revoke Number Verification</Text>
            </TouchableOpacity>
          </View>
          <View style={style.devSettingsItem}>
            <TouchableOpacity onPress={this.resetAppOpenedState}>
              <Text>Reset app opened state</Text>
            </TouchableOpacity>
          </View>
          <View style={style.devSettingsItem}>
            <TouchableOpacity onPress={this.showDebugScreen}>
              <Text>Show Debug Screen</Text>
            </TouchableOpacity>
          </View>
          <View style={style.devSettingsItem}>
            <TouchableOpacity onPress={Sentry.nativeCrash}>
              <Text>Trigger a crash</Text>
            </TouchableOpacity>
          </View>
        </View>
      )
    }
  }

  render() {
    const { t, account } = this.props

    return (
      <ScrollView style={style.scrollView}>
        <SafeAreaView>
          <DisconnectBanner />
          <View style={style.accountProfile}>
            <TouchableOpacity onPress={this.onPressAvatar}>
              <AvatarSelf />
            </TouchableOpacity>
            <View>
              <TouchableOpacity onPress={this.onPressAddress}>
                <Text numberOfLines={1} ellipsizeMode={'tail'} style={style.addressText}>
                  {account}
                </Text>
              </TouchableOpacity>
            </View>
            <SmallButton
              text={t('editProfile')}
              testID={'editProfileButton'}
              onPress={this.goToProfile}
              solid={false}
              style={style.buttonSpacing}
            />
          </View>
          <View style={style.containerList}>
            <SettingsItem title={t('backupKey')} onPress={this.backupScreen} />
            <SettingsItem title={t('invite')} onPress={this.goToInvite} />
            {features.SHOW_SHOW_REWARDS_APP_LINK && (
              <SettingsItem title={t('celoRewards')} onPress={navigateToVerifierApp} />
            )}
            <SettingsItem title={t('analytics')} onPress={this.goToAnalytics} />
            <SettingsItem title={t('languageSettings')} onPress={this.goToLanguageSetting} />
            <SettingsItem
              title={t('localCurrencySetting')}
              onPress={this.goToLocalCurrencySetting}
            />
            <SettingsItem title={t('licenses')} onPress={this.goToLicenses} />
            <SettingsItem title={t('sendIssueReport')} onPress={this.sendLogsToSupport} />
          </View>
          {this.getDevSettingsComp()}

          <View style={style.accountFooter}>
            {DeviceInfo.getVersion() && (
              <View style={style.accountFooterText}>
                <Text style={fontStyles.bodySmall}>
                  {t('version') + ' ' + DeviceInfo.getVersion()}
                </Text>
              </View>
            )}
            <View style={style.accountFooterText}>
              <Trans i18nKey="testFaqHere">
                <Text style={fontStyles.bodySmall}>Test FAQ is </Text>
                <Link style={[fontStyles.bodySmall, fontStyles.linkInline]} onPress={this.goToFAQ}>
                  here
                </Link>
              </Trans>
            </View>
            <View style={style.accountFooterText}>
              <Trans i18nKey="termsOfServiceHere">
                <Text style={fontStyles.bodySmall}>Terms of service are </Text>
                <Link
                  style={[fontStyles.bodySmall, fontStyles.linkInline]}
                  onPress={this.goToTerms}
                >
                  here
                </Link>
              </Trans>
            </View>
          </View>
        </SafeAreaView>
      </ScrollView>
    )
  }
}

const style = StyleSheet.create({
  accountProfile: {
    paddingBottom: 20,
    flexDirection: 'column',
    alignItems: 'center',
  },
  accountFooter: {
    marginTop: 30,
    flexDirection: 'column',
    alignItems: 'center',
  },
  accountFooterText: {
    flexDirection: 'row',
    paddingBottom: 10,
  },
  scrollView: {
    flex: 1,
    backgroundColor: 'white',
  },
  containerList: {
    flex: 1,
    paddingLeft: 20,
    borderTopWidth: 1,
    borderColor: '#EEEEEE',
  },
  devSettings: {
    alignItems: 'flex-start',
    padding: 15,
    marginHorizontal: 10,
  },
  devSettingsItem: {
    alignSelf: 'stretch',
    margin: 4,
  },
  buttonSpacing: {
    marginTop: 10,
    alignSelf: 'center',
  },
  addressText: {
    ...fontStyles.bodySmall,
    ...fontStyles.light,
    color: colors.dark,
    marginTop: 8,
  },
})

export default connect<StateProps, DispatchProps, {}, RootState>(
  mapStateToProps,
  mapDispatchToProps
)(withNamespaces(Namespaces.accountScreen10)(Account))
