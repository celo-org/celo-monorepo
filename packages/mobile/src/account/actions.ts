import { PincodeType } from 'src/account/reducer'
import { PaymentRequest } from 'src/account/types'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { DefaultEventNames } from 'src/analytics/constants'

// TODO(Rossy): Remove the _ACTION suffix from these actions for consistency with other other names
export enum Actions {
  SET_NAME = 'ACCOUNT/SET_NAME',
  SET_PHONE_NUMBER = 'ACCOUNT/SET_PHONE_NUMBER',
  DEV_MODE_TRIGGER_CLICKED = 'ACCOUNT/NAME_CLICKED',
  PHOTOSNUX_CLICKED = 'ACCOUNT/PHOTOSNUX_CLICKED',
  SET_PINCODE = 'ACCOUNT/SET_PINCODE',
  SET_PINCODE_SUCCESS = 'ACCOUNT/SET_PINCODE_SUCCESS',
  SET_PINCODE_FAILURE = 'ACCOUNT/SET_PINCODE_FAILURE',
  SET_ACCOUNT_CREATION_TIME_ACTION = 'ACCOUNT/SET_ACCOUNT_CREATION_TIME_ACTION',
  SET_BACKUP_COMPLETED_ACTION = 'ACCOUNT/SET_BACKUP_COMPLETED_ACTION',
  SET_BACKUP_DELAYED_ACTION = 'ACCOUNT/SET_BACKUP_DELAYED_ACTION',
  SET_SOCIAL_BACKUP_COMPLETED_ACTION = 'ACCOUNT/SET_SOCIAL_BACKUP_COMPLETED_ACTION',
  RESET_BACKUP_STATE = 'ACCOUNT/RESET_BACKUP_STATE',
  UPDATE_INCOMING_PAYMENT_REQUESTS = 'ACCOUNT/UPDATE_INCOMING_PAYMENT_REQUESTS',
  UPDATE_OUTGOING_PAYMENT_REQUESTS = 'ACCOUNT/UPDATE_OUTGOING_PAYMENT_REQUESTS',
  DISMISS_EARN_REWARDS = 'ACCOUNT/DISMISS_EARN_REWARDS',
  DISMISS_INVITE_FRIENDS = 'ACCOUNT/DISMISS_INVITE_FRIENDS',
  SET_USER_CONTACT_DETAILS = 'ACCOUNT/SET_USER_CONTACT_DETAILS',
}

export interface SetNameAction {
  type: Actions.SET_NAME
  name: string
}

export interface SetPhoneNumberAction {
  type: Actions.SET_PHONE_NUMBER
  e164PhoneNumber: string
  countryCode: string
}

export interface DevModeTriggerClickedAction {
  type: Actions.DEV_MODE_TRIGGER_CLICKED
}

export interface PhotosNUXClickedAction {
  type: Actions.PHOTOSNUX_CLICKED
}

export interface SetPincodeAction {
  type: Actions.SET_PINCODE
  pincodeType: PincodeType
  pin?: string
}

export interface SetPincodeSuccessAction {
  type: Actions.SET_PINCODE_SUCCESS
  pincodeType: PincodeType
}

export interface SetPincodeFailureAction {
  type: Actions.SET_PINCODE_FAILURE
}

export interface SetAccountCreationAction {
  type: Actions.SET_ACCOUNT_CREATION_TIME_ACTION
}

export interface SetBackupCompletedAction {
  type: Actions.SET_BACKUP_COMPLETED_ACTION
}

export interface SetBackupDelayedAction {
  type: Actions.SET_BACKUP_DELAYED_ACTION
}

export interface SetSocialBackupCompletedAction {
  type: Actions.SET_SOCIAL_BACKUP_COMPLETED_ACTION
}

export interface ResetBackupState {
  type: Actions.RESET_BACKUP_STATE
}

export interface UpdateIncomingPaymentRequestsAction {
  type: Actions.UPDATE_INCOMING_PAYMENT_REQUESTS
  paymentRequests: PaymentRequest[]
}

export interface UpdateOutgoingPaymentRequestsAction {
  type: Actions.UPDATE_OUTGOING_PAYMENT_REQUESTS
  paymentRequests: PaymentRequest[]
}

export interface DismissEarnRewards {
  type: Actions.DISMISS_EARN_REWARDS
}

export interface DismissInviteFriends {
  type: Actions.DISMISS_INVITE_FRIENDS
}

export interface SetContactDetailsAction {
  type: Actions.SET_USER_CONTACT_DETAILS
  contactId: string
  thumbnailPath: string | null
}

export type ActionTypes =
  | SetNameAction
  | SetPhoneNumberAction
  | DevModeTriggerClickedAction
  | PhotosNUXClickedAction
  | SetPincodeAction
  | SetPincodeSuccessAction
  | SetPincodeFailureAction
  | SetAccountCreationAction
  | SetBackupCompletedAction
  | SetBackupDelayedAction
  | SetSocialBackupCompletedAction
  | ResetBackupState
  | UpdateIncomingPaymentRequestsAction
  | UpdateOutgoingPaymentRequestsAction
  | DismissEarnRewards
  | DismissInviteFriends
  | SetContactDetailsAction

export function setName(name: string): SetNameAction {
  return {
    type: Actions.SET_NAME,
    name,
  }
}

export function setPhoneNumber(e164PhoneNumber: string, countryCode: string): SetPhoneNumberAction {
  CeloAnalytics.track(DefaultEventNames.phoneNumberSet, { countryCode })
  return {
    type: Actions.SET_PHONE_NUMBER,
    e164PhoneNumber,
    countryCode,
  }
}

export const devModeTriggerClicked = (): DevModeTriggerClickedAction => ({
  type: Actions.DEV_MODE_TRIGGER_CLICKED,
})

export const photosNUXCompleted = (): PhotosNUXClickedAction => ({
  type: Actions.PHOTOSNUX_CLICKED,
})

export const setPincode = (pincodeType: PincodeType, pin?: string): SetPincodeAction => ({
  type: Actions.SET_PINCODE,
  pincodeType,
  pin,
})

export const setPincodeSuccess = (pincodeType: PincodeType): SetPincodeSuccessAction => ({
  type: Actions.SET_PINCODE_SUCCESS,
  pincodeType,
})

export const setPincodeFailure = (): SetPincodeFailureAction => ({
  type: Actions.SET_PINCODE_FAILURE,
})

export const setAccountCreationTime = (): SetAccountCreationAction => ({
  type: Actions.SET_ACCOUNT_CREATION_TIME_ACTION,
})

export const setBackupCompleted = (): SetBackupCompletedAction => ({
  type: Actions.SET_BACKUP_COMPLETED_ACTION,
})

export const setBackupDelayed = (): SetBackupDelayedAction => ({
  type: Actions.SET_BACKUP_DELAYED_ACTION,
})

export const setSocialBackupCompleted = (): SetSocialBackupCompletedAction => ({
  type: Actions.SET_SOCIAL_BACKUP_COMPLETED_ACTION,
})

export const resetBackupState = (): ResetBackupState => ({
  type: Actions.RESET_BACKUP_STATE,
})

export const updateIncomingPaymentRequests = (
  paymentRequests: PaymentRequest[]
): UpdateIncomingPaymentRequestsAction => ({
  type: Actions.UPDATE_INCOMING_PAYMENT_REQUESTS,
  paymentRequests,
})

export const updateOutgoingPaymentRequests = (
  paymentRequests: PaymentRequest[]
): UpdateOutgoingPaymentRequestsAction => ({
  type: Actions.UPDATE_OUTGOING_PAYMENT_REQUESTS,
  paymentRequests,
})

export const dismissEarnRewards = (): DismissEarnRewards => ({
  type: Actions.DISMISS_EARN_REWARDS,
})

export const dismissInviteFriends = (): DismissInviteFriends => ({
  type: Actions.DISMISS_INVITE_FRIENDS,
})

export const setUserContactDetails = (
  contactId: string,
  thumbnailPath: string | null
): SetContactDetailsAction => ({
  type: Actions.SET_USER_CONTACT_DETAILS,
  contactId,
  thumbnailPath,
})
