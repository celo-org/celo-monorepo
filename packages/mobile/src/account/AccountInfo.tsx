import ContactCircle from '@celo/react-components/components/ContactCircle'
import PhoneNumberWithFlag from '@celo/react-components/components/PhoneNumberWithFlag'
import QRCode from '@celo/react-components/icons/QRCode'
import { fontStyles } from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import { connect } from 'react-redux'
import { devModeTriggerClicked } from 'src/account/actions'
import { getUserContactDetails, UserContactDetails } from 'src/account/reducer'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { componentWithAnalytics } from 'src/analytics/wrapper'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'

interface DispatchProps {
  devModeTriggerClicked: typeof devModeTriggerClicked
}

interface StateProps {
  name: string
  e164Number: string
  photosNUXClicked: boolean
  userContact: UserContactDetails
  devModeActive: boolean
  defaultCountryCode: string
}

type Props = StateProps & DispatchProps

const mapStateToProps = (state: RootState) => {
  return {
    name: state.account.name,
    e164Number: state.account.e164PhoneNumber,
    photosNUXClicked: state.account.photosNUXClicked,
    userContact: getUserContactDetails(state),
    devModeActive: state.account.devModeActive || false,
    defaultCountryCode: state.account.defaultCountryCode,
  }
}

export class AccountInfo extends React.Component<Props> {
  handleNameClicked = () => {
    this.props.devModeTriggerClicked()
  }

  handlePhotoClicked = () => {
    const { photosNUXClicked } = this.props
    if (!photosNUXClicked) {
      CeloAnalytics.track(CustomEventNames.photos_education)
      navigate(Screens.PhotosEducation)
    } else {
      CeloAnalytics.track(CustomEventNames.qrcode_main_screen_visit)
      navigate(Screens.QRCode)
    }
  }

  render() {
    const { name, e164Number, userContact, defaultCountryCode } = this.props

    return (
      <View style={style.accountProfile}>
        <TouchableOpacity onPress={this.handlePhotoClicked} style={style.photosNUX}>
          <ContactCircle
            thumbnailPath={userContact.thumbnailPath}
            name={name}
            preferNameInitial={true}
            size={55}
          />
          <View style={style.qrcode}>
            <QRCode />
          </View>
        </TouchableOpacity>
        {!!name && (
          <Text
            style={[fontStyles.bodySmallSemiBold, style.nameText]}
            onPress={this.handleNameClicked}
            suppressHighlighting={true}
          >
            {name}
          </Text>
        )}
        {!!e164Number && (
          <PhoneNumberWithFlag
            defaultCountryCode={defaultCountryCode}
            e164PhoneNumber={e164Number}
          />
        )}
      </View>
    )
  }
}

const style = StyleSheet.create({
  accountProfile: {
    alignItems: 'center',
  },
  photosNUX: {
    flexDirection: 'row',
    alignItems: 'flex-end',
    overflow: 'visible',
    padding: 10,
  },
  qrcode: {
    position: 'absolute',
    right: 5,
    bottom: 5,
    zIndex: 10,
  },
  nameText: {
    paddingTop: 10,
  },
  dot: {
    position: 'absolute',
    right: -15,
    bottom: 25,
    zIndex: 20,
  },
})

export default componentWithAnalytics(
  connect<StateProps, DispatchProps, {}, RootState>(
    mapStateToProps,
    { devModeTriggerClicked }
  )(AccountInfo)
)
