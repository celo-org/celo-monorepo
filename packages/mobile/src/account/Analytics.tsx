import SettingsSwitchItem from '@celo/react-components/components/SettingsSwitchItem'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { ScrollView, StyleSheet, Text } from 'react-native'
import { connect } from 'react-redux'
import { setAnalyticsEnabled } from 'src/app/actions'
import i18n, { Namespaces } from 'src/i18n'
import { headerWithBackButton } from 'src/navigator/Headers'
import { RootState } from 'src/redux/reducers'

interface StateProps {
  analyticsEnabled: boolean
}

interface DispatchProps {
  setAnalyticsEnabled: typeof setAnalyticsEnabled
}

type Props = StateProps & DispatchProps & WithNamespaces

const mapStateToProps = (state: RootState): StateProps => {
  return {
    analyticsEnabled: state.app.analyticsEnabled,
  }
}

export class Analytics extends React.Component<Props> {
  static navigationOptions = () => ({
    ...headerWithBackButton,
    headerTitle: i18n.t('accountScreen10:analytics'),
  })

  render() {
    const { analyticsEnabled, t } = this.props
    return (
      <ScrollView style={style.scrollView} keyboardShouldPersistTaps="handled">
        <SettingsSwitchItem
          switchValue={analyticsEnabled}
          onSwitchChange={this.props.setAnalyticsEnabled}
          details={t('shareAnalytics_detail')}
        >
          <Text style={fontStyles.body}>{t('shareAnalytics')}</Text>
        </SettingsSwitchItem>
      </ScrollView>
    )
  }
}

const style = StyleSheet.create({
  scrollView: {
    flex: 1,
    backgroundColor: colors.background,
  },
})

export default connect<StateProps, DispatchProps, {}, RootState>(
  mapStateToProps,
  { setAnalyticsEnabled }
)(withNamespaces(Namespaces.accountScreen10)(Analytics))
