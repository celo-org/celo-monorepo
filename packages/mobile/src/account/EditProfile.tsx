import TextInput from '@celo/react-components/components/TextInput'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { ScrollView, StyleSheet } from 'react-native'
import { connect } from 'react-redux'
import { setName } from 'src/account/actions'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { Namespaces, withTranslation } from 'src/i18n'
import { headerWithBackButton } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'

interface StateProps {
  name: string
}

interface DispatchProps {
  setName: typeof setName
}

type Props = StateProps & DispatchProps & WithTranslation

const mapStateToProps = (state: RootState): StateProps => {
  return {
    name: state.account.name,
  }
}

export class EditProfile extends React.Component<Props> {
  static navigationOptions = headerWithBackButton

  state = {
    name: this.props.name,
  }

  nameValueChange = (name: string) => {
    this.setState({ name })
  }

  onEndEditing = () => {
    CeloAnalytics.track(CustomEventNames.edit_name_input)
  }

  nameSubmitted = () => {
    this.props.setName(this.state.name)
    CeloAnalytics.track(CustomEventNames.edit_name_submit)
    navigate(Screens.Account)
  }

  render() {
    const { t } = this.props
    return (
      <ScrollView style={style.scrollView} keyboardShouldPersistTaps="handled">
        <TextInput
          style={[style.input, fontStyles.regular]}
          underlineColorAndroid={'transparent'}
          autoFocus={true}
          autoCorrect={false}
          placeholder={t('yourName')}
          value={this.state.name}
          onSubmitEditing={this.nameSubmitted}
          onChangeText={this.nameValueChange}
          onEndEditing={this.onEndEditing}
        />
      </ScrollView>
    )
  }
}

const style = StyleSheet.create({
  input: {
    borderWidth: 1,
    borderRadius: 3,
    borderColor: '#EEEEEE',
    padding: 5,
    height: 54,
    margin: 20,
    width: variables.width - 40,
    fontSize: 16,
  },
  scrollView: {
    flex: 1,
    backgroundColor: colors.background,
  },
})

export default connect<StateProps, DispatchProps, {}, RootState>(mapStateToProps, {
  setName,
})(withTranslation(Namespaces.accountScreen10)(EditProfile))
