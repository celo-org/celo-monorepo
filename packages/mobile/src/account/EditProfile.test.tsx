import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import { setName } from 'src/account/actions'
import { EditProfile } from 'src/account/EditProfile'
import { createMockStore, getMockI18nProps } from 'test/utils'

jest.mock('src/web3/contracts', () => ({
  isZeroSyncMode: jest.fn().mockReturnValueOnce(false),
}))

it('renders the EditProfile Component', () => {
  const store = createMockStore()
  const tree = renderer.create(
    <Provider store={store}>
      <EditProfile {...getMockI18nProps()} name={'Test'} setName={setName} />
    </Provider>
  )
  expect(tree).toMatchSnapshot()
})
