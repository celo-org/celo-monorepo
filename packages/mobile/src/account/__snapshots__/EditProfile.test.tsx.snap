// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the EditProfile Component 1`] = `
<RCTScrollView
  keyboardShouldPersistTaps="handled"
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          Array [
            Object {
              "borderColor": "#EEEEEE",
              "borderRadius": 3,
              "borderWidth": 1,
              "fontSize": 16,
              "height": 54,
              "margin": 20,
              "padding": 5,
              "width": 710,
            },
            Object {
              "fontFamily": "Hind-Regular",
            },
          ],
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCorrect={false}
        autoFocus={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onEndEditing={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="yourName"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#D1D5D8",
            "borderRadius": 3,
            "flex": 1,
            "fontFamily": "Hind-Regular",
            "padding": 8,
          }
        }
        underlineColorAndroid="transparent"
        value="Test"
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "marginRight": 8,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(0, 0, 0, .4)",
              "borderColor": "rgba(0, 0, 0, .4)",
              "borderRadius": 10,
              "borderWidth": 0,
              "height": 20,
              "justifyContent": "center",
              "opacity": 1,
              "width": 20,
            }
          }
        >
          <svg
            fill="none"
            height={8}
            style={Object {}}
            viewBox="0 0 92.132 92.132"
            width={8}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.141,89.13c1.425,1.429,3.299,2.142,5.167,2.142c1.869,0,3.742-0.713,5.167-2.142l33.591-33.592L79.657,89.13
          c1.426,1.429,3.299,2.142,5.167,2.142c1.867,0,3.74-0.713,5.167-2.142c2.854-2.854,2.854-7.48,0-10.334L56.398,45.205
          l31.869-31.869c2.855-2.853,2.855-7.481,0-10.334c-2.853-2.855-7.479-2.855-10.334,0L46.065,34.87L14.198,3.001
          c-2.854-2.855-7.481-2.855-10.333,0c-2.855,2.853-2.855,7.481,0,10.334l31.868,31.869L2.143,78.795
          C-0.714,81.648-0.714,86.274,2.141,89.13z"
              fill="#FFFFFF"
              style={Object {}}
            />
          </svg>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
